core/src/main/java/org/ldaptive/url/DefaultUrlParser.java

```diff
@@ -0,0 +1,232 @@
+      currentPos = nextPos;
+      nextPos = url.indexOf('?', currentPos + 1);
+      attrs = nextPos < 0 ? parseAttributes(url.substring(currentPos + 1)) :
+        parseAttributes(url.substring(currentPos + 1, nextPos));
+    }
+
+    SearchScope scope = null;
+    if (nextPos > -1) {
+      currentPos = nextPos;
+      nextPos = url.indexOf('?', currentPos + 1);
+      scope = nextPos < 0 ? parseScope(url.substring(currentPos + 1)) :
+        parseScope(url.substring(currentPos + 1, nextPos));
+    }
+
+    String filter = null;
+    Filter parsedFilter = null;
+    if (nextPos > -1) {
+      try {
+        filter = LdapUtils.percentDecode(url.substring(nextPos + 1));
+        if (filter.isEmpty()) {
+          filter = null;
+        } else {
+          parsedFilter = filterFunction.parse(filter);
+        }
+      } catch (Exception e) {
+        throw new IllegalArgumentException("Invalid filter: " + url, e);
+      }
+    }
+
+    return new Url(scheme, hostname, port, baseDn, parsedBaseDn, attrs, scope, filter, parsedFilter);
+  }
+
+
+  /**
+   * Parses the supplied string containing a colon delimited hostname and port.
+   *
+   * @param  hostAndPort  to parse
+   *
+   * @return  array containing a string hostname in the first index and an integer port in the second index
+   */
+  private static Object[] parseHostAndPort(final String hostAndPort)
+  {
+    if (hostAndPort == null) {
+      throw new IllegalArgumentException("Host and port cannot be empty or null");
+    }
+    if (hostAndPort.isEmpty()) {
+      return new Object[] {null, -1};
+    }
+
+    // check for IPv6 address
+    if (hostAndPort.indexOf('[') == 0) {
+      final int bracketPos = hostAndPort.indexOf(']');
+      if (bracketPos < 0) {
+        throw new IllegalArgumentException("IPv6 address is missing closing bracket: " + hostAndPort);
+      }
+      final String ipv6Address = LdapUtils.trimSpace(hostAndPort.substring(1, bracketPos));
+      if (ipv6Address.isEmpty()) {
+        throw new IllegalArgumentException("IPv6 address is empty: " + hostAndPort);
+      }
+      final int ipv6Port;
+      if (bracketPos == hostAndPort.length() - 1) {
+        ipv6Port = -1;
+      } else if (hostAndPort.charAt(bracketPos + 1) != ':') {
+        throw new IllegalArgumentException("IPv6 address invalid port designation: " + hostAndPort);
+      } else {
+        ipv6Port = Integer.parseInt(hostAndPort.substring(bracketPos + 2));
+        Url.validatePort(ipv6Port, false);
+      }
+      return new Object[] {ipv6Address, ipv6Port};
+    }
+
+    final int colonPos = hostAndPort.indexOf(':');
+    final String hostname;
+    final int port;
+    if (colonPos < 0) {
+      hostname = hostAndPort;
+      port = -1;
+    } else {
+      hostname = hostAndPort.substring(0, colonPos);
+      port = Integer.parseInt(hostAndPort.substring(colonPos + 1));
+      Url.validatePort(port, false);
+    }
+    return new Object[] {hostname.isEmpty() ? null : hostname, port};
```

Comment:
Seems like you should throw here rather than returning null. I can't imagine a URL without a hostname is a valid condition.

Reply:
I made it valid for the `UrlParser` implementation, but invalid for `LdapURL`. Technically host is not required: https://datatracker.ietf.org/doc/html/rfc4516#section-3

Reply:
I just realized that you already have a class, `LdapURL` to describe LDAP URLs. You ought to have exactly one class for this concern in the library and use it consistently. I can see that `LdapURL` perhaps needs to continue to exist as an API bridge, but you should move to consolidate on a single data model. Then I will buy the argument that it can be null, as it will be nullable everywhere.

Reply:
One goal of this patch is to separate the concerns of LDAP connections from URL parsing. The `url` package was created to address the latter. The `LdapURL` class was refactored to address the former. Since `LdapURL` contains state data specific to a TCP connection and is produced by `ConnectionStrategy` it made more sense for it to have an instance of an parsed URL to use for its purposes.

Meanwhile, any use cases that require URL parsing can use the `url` package.

Reply:
I encourage you to consider consolidation at some point in the future.###